pkg/providers/prometheus.go

```diff
@@ -122,80 +136,124 @@ func (l *PrometheusMP) Update() error {
+
+	var v model.Value
 	select {
-	case value := <-ch:
-		cancelFunc()
-		return value
+	case v = <-valueCh:
 	case <-ctx.Done():
-		return nil
+		l.log.Error(nil, "all requests failed")
 	}
+	return v
 }
 
-func (l *PrometheusMP) Load(production map[int32]int64, consumption map[int32]int64, offset map[int32]int64) {
-	l.productionRate = production
-	l.consumptionRate = consumption
-	l.messagesBehind = offset
-}
-
-func (l *PrometheusMP) queryOffset() (MetricsMap, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), promCallTimeout)
-	defer cancel()
-	ch := make(chan model.Value)
-	value := l.queryAll(ctx, l.offsetQuery, ch, cancel)
+func (l *PrometheusMP) queryOffset() (metricsMap, error) {
+	value := l.queryAll(l.offsetQuery)
 	if value == nil {
 		return nil, errors.New("failed to get offset metrics from prometheus")
 	}
-	metrics := l.parseVector(value, l.offsetPartitionLabel)
+	metrics := l.parse(value, l.offsetPartitionLabel)
 	return metrics, nil
-
 }
 
 // queryProductionRate queries Prometheus for the current production rate
-func (l *PrometheusMP) queryProductionRate() (MetricsMap, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), promCallTimeout)
-	defer cancel()
-	ch := make(chan model.Value)
-	value := l.queryAll(ctx, l.productionQuery, ch, cancel)
+func (l *PrometheusMP) queryProductionRate() (metricsMap, error) {
+	value := l.queryAll(l.productionQuery)
 	if value == nil {
 		return nil, errors.New("failed to get production metrics from prometheus")
 	}
-	metrics := l.parseVector(value, l.productionPartitionLabel)
+	metrics := l.parse(value, l.productionPartitionLabel)
 	return metrics, nil
 }
 
-func (l *PrometheusMP) queryConsumptionRate() (MetricsMap, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), promCallTimeout)
-	defer cancel()
-	ch := make(chan model.Value)
-	value := l.queryAll(ctx, l.consumptinQuery, ch, cancel)
+func (l *PrometheusMP) queryConsumptionRate() (metricsMap, error) {
+	value := l.queryAll(l.consumptionQuery)
 	if value == nil {
-		return nil, errors.New("failed to get production metrics from prometheus")
+		return nil, errors.New("failed to get consumption metrics from prometheus")
 	}
-	metrics := l.parseVector(value, l.consumptionPartitionLabel)
+	metrics := l.parse(value, l.consumptionPartitionLabel)
 	return metrics, nil
 }
 
-func (l *PrometheusMP) parseVector(v model.Value, lbl string) MetricsMap {
-	metrics := make(MetricsMap)
-	// TODO: cover cases when value is not a vector
-	for _, v := range v.(model.Vector) {
-		partitionNumberStr := string(v.Metric[model.LabelName(l.productionPartitionLabel)])
+func (l *PrometheusMP) parse(value model.Value, label model.LabelName) metricsMap {
+	switch value.Type() {
+	case model.ValVector:
+		return l.parseVector(value.(model.Vector), label)
+	case model.ValMatrix:
+		return l.parseMatrix(value.(model.Matrix), label)
+	default:
+		panic(fmt.Sprintf("unsupported prometheus response type %q", value.Type()))
```

Comment:
Are you sure we should ever panic in the controller?

Reply:
removed